README.md
# TatSuRDG
A random derivation generator for grammars that were compiled with the TatSu PEG/Packrat parser generator for Python.

A typical application is testing grammars by generating some derivations and checking if they look like intended. 

It is also possible to replace subrules of the grammar with constant strings, so the derivations become more simple and more modular. 

Regex patterns are derived using the rstr package.

CHANGES.md
# 1.0.0
* Initial commit
# 1.0.1
* Bug fixes
# 1.0.2
* Replace rule derivations by placeholders
* README.md added
* Code commented
# 1.0.3
* CHANGES.md added
* Optional trace argument added that produces a list of named grammar rules that were visited during the derivation process.

tests/test_calc_factored.py
import unittest
from parameterized import parameterized
from tatsurd import TatSuRDG
import os
import io
import tatsu


class TatSuRDTests(unittest.TestCase):
    path = ""
    path_to_grammar = ""
    path_to_usecases = ""
    parser = None
    rd = None

    @classmethod
    def setUpClass(cls):
        cls.path = os.path.normpath(os.path.abspath(__file__))
        if os.path.isfile(cls.path):
            cls.path = os.path.dirname(cls.path)
        cls.path_to_grammar = os.path.join(cls.path, "calc_factored.ebnf")
        cls.path_to_usecases = os.path.join(cls.path)
        with io.open(cls.path_to_grammar, 'r', encoding="utf-8") as file:
            grammar = file.read()
            cls.parser = tatsu.compile(grammar)
            cls.rg = TatSuRDG(cls.parser, max_length_regex=5, max_counter=5, recursion_limit=1500)

    @parameterized.expand([
        "expression",
        "term",
        "+",
        "-",
        "*",
        "/",
        "factor",
        "(",
        ")",
        "number",
    ])
    def test_derivation_generator(self, use_case):
        for trial in range(1, 50):
            self.rg.init_rules()
            self.rg.random_derivation(use_case)


tests/test_fpl_generator.py
import unittest
from parameterized import parameterized
from tatsurd import TatSuRDG
import os
import io
import tatsu


class TatSuRDTests(unittest.TestCase):
    path = ""
    path_to_grammar = ""
    path_to_usecases = ""
    parser = None
    rd = None

    @classmethod
    def setUpClass(cls):
        cls.path = os.path.normpath(os.path.abspath(__file__))
        if os.path.isfile(cls.path):
            cls.path = os.path.dirname(cls.path)
        cls.path_to_grammar = os.path.join(cls.path, "fpl.ebnf")
        cls.path_to_usecases = os.path.join(cls.path)
        with io.open(cls.path_to_grammar, 'r', encoding="utf-8") as file:
            grammar = file.read()
            cls.parser = tatsu.compile(grammar)
            cls.rg = TatSuRDG(cls.parser, max_length_regex=5, max_counter=5, recursion_limit=1500)

    @parameterized.expand([
        "Namespace",
        "NamespaceBlock",
        "WildcardTheoryNamespaceList",
        "ExtensionBlock",
        "ExtensionHeader",
        "ExtensionContent",
        "ExtensionTail",
        "ext",
        "end",
        "UsesClause",
        "TheoryBlock",
        "RulesOfInferenceBlock",
        "InferenceHeader",
        "inference",
        "inf",
        "PremiseHeader",
        "premise",
        "pre",
        "ConclusionHeader",
        "conclusion",
        "con",
        "RuleOfInferenceList",
        "RuleOfInference",
        "PremiseConclusionBlock",
        "PremiseBlock",
        "ConclusionBlock",
        "TheoryHeader",
        "theory",
        "th",
        "BuildingBlockList",
        "BuildingBlock",
        "uses",
        "Definition",
        "DefinitionClass",
        "ObjectDefinitionBlock",
        "FunctionalTermDefinitionBlock",
        "DefinitionContentList",
        "DefinitionContent",
        "DefinitionFunctionalTerm",
        "FunctionalTermSignature",
        "Mapping",
        "VariableSpecificationList",
        "VariableSpecification",
        "AssertionStatement",
        "assert",
        "NamedVariableDeclaration",
        "VariableType",
        "GeneralType",
        "IndexHeader",
        "index",
        "ind",
        "Type",
        "VariableList",
        "DefinitionPredicate",
        "PredicateHeader",
        "FunctionalTermHeader",
        "PredicateDefinitionBlock",
        "PredicateIdentifier",
        "AliasedId",
        "ClassHeader",
        "class",
        "cl",
        "predicate",
        "pred",
        "function",
        "func",
        "To",
        "ObjectHeader",
        "LongTemplateHeader",
        "TemplateHeader",
        "object",
        "obj",
        "template",
        "tpl",
        "Axiom",
        "AxiomHeader",
        "AxiomBlock",
        "axiom",
        "ax",
        "postulate",
        "post",
        "PropertyList",
        "Property",
        "DefinitionProperty",
        "ClassInstance",
        "FunctionalTermInstance",
        "PropertyHeader",
        "mandatory",
        "mand",
        "optional",
        "opt",
        "Constructor",
        "InstanceBlock",
        "Signature",
        "ParenthesisedGeneralType",
        "ParamTuple",
        "ArgumentParam",
        "NamedVariableDeclarationList",
        "TypeWithCoord",
        "EntityWithCoord",
        "CoordList",
        "CoordInSignature",
        "Entity",
        "AtList",
        "Coord",
        "RangeInSignature",
        "Range",
        "CallModifier",
        "Plus",
        "ConditionFollowedByResultList",
        "ConditionFollowedByResult",
        "At",
        "TheoremLikeStatementOrConjecture",
        "TheoremLikeStatementOrConjectureHeader",
        "theorem",
        "thm",
        "proposition",
        "prop",
        "lemma",
        "lem",
        "corollary",
        "cor",
        "conjecture",
        "conj",
        "Proof",
        "ProofIdentifier",
        "ProofHeadHeader",
        "ProofBlock",
        "ProofArgumentList",
        "ProofArgument",
        "VDash",
        "AssumeHeader",
        "AssumedPredicate",
        "DerivedPredicate",
        "qed",
        "trivial",
        "PremiseOrOtherPredicate",
        "ArgumentInference",
        "Revoke",
        "RevokeHeader",
        "Justification",
        "Slash",
        "proof",
        "prf",
        "assume",
        "ass",
        "revoke",
        "rev",
        "true",
        "false",
        "not",
        "and",
        "or",
        "impl",
        "iif",
        "xor",
        "is",
        "Predicate",
        "PrimePredicate",
        "QualifiedIdentifier",
        "PredicateWithArguments",
        "Identifier",
        "AmpersandVariable",
        "IndexValue",
        "Ampersand",
        "self",
        "PredicateList",
        "CompoundPredicate",
        "IsOperator",
        "Negation",
        "Conjunction",
        "Disjunction",
        "Implication",
        "Equivalence",
        "ExclusiveOr",
        "ParenthesisedPredicate",
        "All",
        "all",
        "Exists",
        "ex",
        "ExistsHeader",
        "ExistsTimesN",
        "Dollar",
        "PythonDelegate",
        "PythonIdentifier",
        "py",
        "LeftBracket",
        "RightBracket",
        "StatementList",
        "Statement",
        "ReturnStatement",
        "CaseStatement",
        "LoopStatement",
        "RangeStatement",
        "RangeOrLoopBody",
        "KeysOfVariadicVariable",
        "ClosedOrOpenRange",
        "LeftBound",
        "RightBound",
        "ExclamationMark",
        "range",
        "loop",
        "DefaultResult",
        "AssignmentStatement",
        "Assignee",
        "ReturnHeader",
        "UndefinedHeader",
        "return",
        "ret",
        "undefined",
        "undef",
        "ColonEqual",
        "else",
        "case",
        "Dot",
        "LeftBrace",
        "RightBrace",
        "LeftParen",
        "RightParen",
        "Comma",
        "Star",
        "Colon",
        "NamespaceIdentifier",
        "ArgumentIdentifier",
        "WildcardTheoryNamespace",
        "NamespaceModifier",
        "Alias",
        "alias",
        "IdStartsWithCap",
        "IdStartsWithSmallCase",
        "SW",
        "CW",
        "Comment",
        "LongComment",
        "IW",
        "XId",
        "Variable",
        "Digit",
        "LocalizationBlock",
        "LocalizationList",
        "Localization",
        "SemiColon",
        "Tilde",
        "TranslationList",
        "Translation",
        "LanguageCode",
        "EBNFTransl",
        "EBNFTerm",
        "EBNFFactor",
        "EBNFString",
        "EBNFBar",
        "LocalizationHeader",
        "localization",
        "loc",
        "extDigit"
    ])
    def test_derivation_generator(self, use_case):
        for trial in range(1, 3):
            self.rg.init_rules()
            self.rg.random_derivation(use_case)


tests/test_select_choice.py
from tatsurd import TatSuRDG
import os
import io
import tatsu
import unittest
import scipy.stats as stats


class TatSuRDTests(unittest.TestCase):
    path = ""
    path_to_grammar = ""
    path_to_usecases = ""
    parser = None
    rd = None

    @classmethod
    def setUpClass(cls):
        cls.path = os.path.normpath(os.path.abspath(__file__))
        if os.path.isfile(cls.path):
            cls.path = os.path.dirname(cls.path)
        cls.path_to_grammar = os.path.join(cls.path, "calc.ebnf")
        cls.path_to_usecases = os.path.join(cls.path)
        with io.open(cls.path_to_grammar, 'r', encoding="utf-8") as file:
            grammar = file.read()
            cls.parser = tatsu.compile(grammar)
            cls.rg = TatSuRDG(cls.parser, max_length_regex=10, max_counter=5, recursion_limit=1500)

    def test_select_choice_n(self):
        counters = dict()
        numb = 10
        for j in range(0, numb):
            counters[j] = 0
        total = 1000000
        for i in range(1, total):
            res = self.rg.select_index_randomly(numb)
            if res is not None:
                counters[res] += 1
        expected = []
        observed = []
        for j in range(0, numb):
            expected.append(total * (j + 1) / (numb * (numb + 1) / 2))
            observed.append(counters[j])

        # normalize the data such that the observed and expected frequencies have the same sum
        again = True
        while again:
            try:
                for j in range(0, numb):
                    observed[j] = observed[j] * sum(expected) / sum(observed)
                result = stats.chisquare(f_obs=observed, f_exp=expected)
                again = False
            except ValueError:
                pass

        print("expected: " + str(expected) + ", sum = " + str(sum(expected)))
        print("observed: " + str(observed) + ", sum = " + str(sum(observed)))
        result = stats.chisquare(f_obs=observed, f_exp=expected)
        print(result)
        # fail if chi square pvalue is smaller than 5%
        self.assertGreater(result[1], 0.05)


tatsurd.py
import tatsu.grammars
import random
import sys
import rstr


class RuleInfo:
    def __init__(self):
        self.max_counter = 0
        self.visit_counter = 0
        self.grammar = None


class TatSuRDG:
    """
    The class TatSuRDG provides a functionality to generate derivations of grammars
    compiled into a TatSu parser randomly
    """
    version = "1.0.2"

    def __init__(self, parser, max_length_regex=5, max_counter=5, recursion_limit=1000, override_placeholders=dict()):
        """
        :param parser: a compiled TatSu parser for some grammar
        :param max_length_regex: TatSuRDG will try to limit the derivations for all terminals of the grammar with a regex patterns to this length
        :param max_counter: While generating a derivation, TatSuRDG might 'visit' a choice rule (a|b|..) multiple times. TatSuRDG will select all options randomly until each has been chosen max_counter of times. From then on, TatSuRDG will select only the LAST option. Since TatSu parsers are PEG parsers, the last option is the 'most simple' one. max_counter prevents infinite derivations or derivations becoming too complex.
        :param recursion_limit: If you encounter RecursionError try to increase the default value for complex grammars.
        :param override_placeholders: Overrides the default derivations of single rules by deriving a placeholder for them. The dictionary is a key value list where the key is the name of the rule and the value is a string representing the placeholder.
        """
        self.parser = parser
        # dict of rules whose derivations will be replaced by placeholder values
        self.override_placeholders = override_placeholders
        self.max_counter = max_counter
        self.choice_rules = dict()
        self._visited = set()
        self.counters = dict()
        self.init_rules()
        self.max_length_regex = max_length_regex
        sys.setrecursionlimit(recursion_limit)
        print("TatSu Random Derivation Generator ({0}) initialized with {1}".format(self.version,
                                                                                    str(sys.getrecursionlimit())))

    def random_derivation(self, rule, trace=None):
        self._visit(rule, trace)
        if str(rule) in self.override_placeholders:
            return str(self.override_placeholders[str(rule)])
        if type(rule) is not str:
            # we have a tatsu grammars object
            tatsu_rule_object = rule
        else:
            # we have the name of the rule, for which we look for the tatsu grammar object
            if rule in self.parser.rulemap:
                tatsu_rule_object = self.parser.rulemap[rule]
            else:
                # e.g. 'something'
                return rule
        if type(tatsu_rule_object) is tatsu.grammars.Sequence:
            # e.g. "x y z"
            ret = ""
            for sequence_elem in tatsu_rule_object.sequence:
                ret += self.random_derivation(sequence_elem, trace)
            return ret
        elif type(tatsu_rule_object) is tatsu.grammars.Pattern:
            # e.g. "/\d+/"
            r = rstr.xeger(rule.pattern)
            counter = 0
            while len(r) > self.max_length_regex and counter < 1000:
                counter += 1
                r = rstr.xeger(rule.pattern)
            if counter >= 1000:
                raise ValueError(
                    "Cannot derive instance of regex /{0}/ that is shorter than {1}".format(str(rule.pattern),
                                                                                            self.max_length_regex))
            else:
                return r
        elif type(tatsu_rule_object) is tatsu.grammars.Token:
            # e.g. "x" -> terminal !
            return tatsu_rule_object.token
        elif type(tatsu_rule_object) is tatsu.grammars.Constant:
            # e.g. "'TATSU'" -> terminal !
            return str(tatsu_rule_object.ast)
        elif type(tatsu_rule_object) is tatsu.grammars.Group:
            # e.g. (x)
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Choice:
            # e.g. (x | y)
            option = self.select_choice(str(tatsu_rule_object))
            return self.random_derivation(option, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.PositiveClosure:
            # e.g. { x }+ or
            r = random.randint(1, 3)
            ret = ""
            while r > 0:
                ret += self.random_derivation(tatsu_rule_object.exp, trace)
                r -= 1
            return ret
        elif type(tatsu_rule_object) is tatsu.grammars.Named:
            # e.g. some_name:{x}+ or some_name:{x}*
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Closure:
            # e.g. { x }*
            r = random.randint(0, 2)
            ret = ""
            while r > 0:
                ret += self.random_derivation(tatsu_rule_object.exp, trace)
                r -= 1
            return ret
        elif type(tatsu_rule_object) is tatsu.grammars.PositiveGather:
            # e.g. s.{ x }+
            r = random.randint(1, 3)
            ret = ""
            while r > 0:
                ret += self.random_derivation(tatsu_rule_object.JOINOP, trace)
                ret += self.random_derivation(tatsu_rule_object.exp, trace)
                r -= 1
            return ret
        elif type(tatsu_rule_object) is tatsu.grammars.RuleRef:
            # e.g. x
            # return the expansion of the referenced rule
            return self.random_derivation(tatsu_rule_object.ast, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Rule:
            # e.g. x
            # return the expansion of the rule
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.NamedList:
            # e.g. name+:x
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.NegativeLookahead:
            # e.g. !x
            # TatSu parser will fail if x can be parsed but does not consume any input
            # we want derivations anyway
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Lookahead:
            # e.g. &x will succeed if x can be parsed but does not consume any input
            # we want derivations anyway
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Cut:
            # e.g. ~
            return ""  # ignore, since we are deriving already the option, other options are cut anyway
        elif type(tatsu_rule_object) is tatsu.grammars.Override:
            # e.g. y = '(' @:x ')' (makes the AST for y to be the AST for x
            # ignore, since we are interested not in the AST but in the syntactical derivations
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.OverrideList:
            # e.g. y = '(' @+:x ')' (makes the AST for y to be the AST for x and forces it to be a list)
            # ignore, since we are interested not in the AST but in the syntactical derivations
            return self.random_derivation(tatsu_rule_object.exp, trace)
        elif type(tatsu_rule_object) is tatsu.grammars.Optional:
            # e.g. [x]
            r = random.randint(0, 1)
            if r > 0:
                # chance 50% the option will be derived
                return self.random_derivation(tatsu_rule_object.exp, trace)
            else:
                return ""
        elif type(tatsu_rule_object) is tatsu.grammars.Void:
            # e.g. ()
            return ""
        elif type(tatsu_rule_object) is tatsu.grammars.EOF:
            # $
            return ""
        elif type(tatsu_rule_object) is tatsu.grammars.RuleInclude:
            # e.g. >x
            # includable = exp1 ; expanded = exp0 >includable exp2; has the same effect as defining expanded as:
            # expanded = exp0 exp1 exp2;
            return self.random_derivation(tatsu_rule_object.exp, trace)
        else:
            raise TypeError(str(type(tatsu_rule_object)))

    def _visit(self, tatsu_rule_object, trace=None):
        self.counters[str(tatsu_rule_object)].visit_counter += 1
        if trace is not None:
            if hasattr(tatsu_rule_object, "name"):
                trace.append(tatsu_rule_object.name)

    def init_rules(self):
        self._visited.clear()
        for rule in self.parser.rulemap:
            self._init_rules_rek(rule)

    def _init_rules_rek(self, rule):
        rule_found = True
        if type(rule) is not str:
            # we have a tatsu grammars object
            tatsu_rule_object = rule
        else:
            if rule in self.parser.rulemap:
                # we have the name of the rule, for which we look for the tatsu grammar object
                tatsu_rule_object = self.parser.rulemap[rule]
            else:
                # the string was a TatSu constant or token.ast
                tatsu_rule_object = rule
                rule_found = False
        if str(tatsu_rule_object) in self._visited:
            return
        else:
            rule_name = str(rule)
            self._visited.add(rule_name)
            ri = RuleInfo()
            ri.max_counter = self.max_counter
            ri.grammar = tatsu_rule_object
            self.counters[rule_name] = ri
            if type(tatsu_rule_object) is tatsu.grammars.Constant or type(tatsu_rule_object) is str:
                # do not resolve constants any more
                return
        if rule_found:
            if type(tatsu_rule_object) is tatsu.grammars.Sequence:
                for sequence_elem in tatsu_rule_object.sequence:
                    self._init_rules_rek(sequence_elem)
            elif type(tatsu_rule_object) is tatsu.grammars.Pattern:
                return  # not a choice, no more recursive rules
            elif type(tatsu_rule_object) is tatsu.grammars.Token:
                self._init_rules_rek(tatsu_rule_object.ast)
            elif type(tatsu_rule_object) is tatsu.grammars.Group:
                # e.g. (x)
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Choice:
                # e.g. (x | y)
                # choice found
                self.choice_rules[str(rule)] = rule
                for option in tatsu_rule_object.options:
                    self._init_rules_rek(option)
            elif type(tatsu_rule_object) is tatsu.grammars.PositiveClosure:
                # e.g. { x }+
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.PositiveGather:
                # e.g. s.{ x }+
                self._init_rules_rek(tatsu_rule_object.JOINOP)  # register s
                self._init_rules_rek(tatsu_rule_object.exp)  # register x
            elif type(tatsu_rule_object) is tatsu.grammars.Named:
                # e.g. some_name:{x}+
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.RuleRef:
                # e.g. x
                # return the expansion of the referenced rule
                self._init_rules_rek(tatsu_rule_object.ast)
            elif type(tatsu_rule_object) is tatsu.grammars.Rule:
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.EOF:
                return  # not a choice, no more recursive rules
            elif type(tatsu_rule_object) is tatsu.grammars.Closure:
                # e.g. { x }*
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Constant:
                # e.g. 'TATSU'
                self._init_rules_rek(tatsu_rule_object.ast)
            elif type(tatsu_rule_object) is tatsu.grammars.NamedList:
                # e.g. name+:x
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.NegativeLookahead:
                # e.g. !x
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Lookahead:
                # e.g. &x
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Cut:
                # e.g. ~
                return
            elif type(tatsu_rule_object) is tatsu.grammars.Override:
                # e.g. y = '(' @:x ')' (makes the AST for y to be the AST for x)
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.OverrideList:
                # e.g. y = '(' @+:x ')' (makes the AST for y to be the AST for x and forces it to be a list)
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Optional:
                # e.g. [x]
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.RuleInclude:
                # e.g. >x
                # includable = exp1 ; expanded = exp0 >includable exp2; has the same effect as defining expanded as:
                # expanded = exp0 exp1 exp2;
                self._init_rules_rek(tatsu_rule_object.exp)
            elif type(tatsu_rule_object) is tatsu.grammars.Void:
                # e.g. ()
                return
            else:
                raise TypeError(str(type(tatsu_rule_object)))
        else:
            # done
            pass

    def select_choice(self, rule_name):
        options = self._get_unvisited_options(rule_name)
        n = len(options)
        if n == 0:
            # if no more unvisited options (all options have been visited at least once)
            # return the last option. Since TatSu is a PEG parser, the last option is the "most simple one"
            return self.choice_rules[rule_name].options[-1]
        else:
            # select an option randomly based on the number of options, preferring the later ones
            return options[self.select_index_randomly(n)]

    def _get_unvisited_options(self, rule_name):
        options = []
        for option in self.choice_rules[rule_name].options:
            option_name = str(option)
            ri = self.counters[option_name]
            if ri.visit_counter < ri.max_counter:
                # add the option only if it was not yet visited the maximal times
                options.append(option)
        return options

    @staticmethod
    def select_index_randomly(n: int):
        if type(n) is not int:
            raise ValueError(str(n) + " must be a positive integer")
        elif n <= 0:
            raise ValueError(str(n) + " must be a positive integer")
        r = random.randint(1, n * (n + 1) // 2)
        i = 0
        total = 0
        while i <= n:
            i += 1
            total += i
            if total >= r:
                break
        if i > n:
            return None
        elif total >= r:
            return i - 1
        else:
            return None

    def _identify_options(self):
        ret = dict()
        for rule in self.parser.rulemap:
            if type(rule) is tatsu.grammars.Choice:
                ret[str(rule)] = rule
        return ret


example.py
import io
import tatsu
from tatsurd import TatSuRDG


class ExampleRandomDerivation:

    def __init__(self, grammar_name, override_placeholders: dict):
        path_to_grammar = "tests/" + grammar_name
        with io.open(path_to_grammar, 'r', encoding="utf-8") as file:
            grammar = file.read()
        self.parser = tatsu.compile(grammar)
        self.rg = TatSuRDG(self.parser, max_length_regex=5, max_counter=5, recursion_limit=1500,
                           override_placeholders=override_placeholders)

    def derive(self, rule_name, trace):
        return self.rg.random_derivation(rule_name, trace)


ex = ExampleRandomDerivation("fpl.ebnf", {"IW": "", "LongComment": "", "Comment": ""})
for i in range(1, 50):
    ex.rg.init_rules()
    trace = []
    print(ex.derive("PredicateWithArguments", trace))
    print(str(trace))
